lib/hasCycle.js

```diff
@@ -8,26 +8,31 @@ const getNextJobs = require('./getNextJobs');
  * @param  {Object} workflowGraph   Directed graph representation of workflow
  * @param  {String} jobName         Job Name
  * @param  {Set} visitedJobs        Unique list of visited jobs
+ * @param  {Array} path             Recursive stack of a single path from node to leaf
  * @return {Boolean}                True if a cycle detected
  * @private
  */
-const walk = (workflowGraph, jobName, visitedJobs) => {
+const isCyclic = (workflowGraph, jobName, visitedJobs, path) => {
```

Comment:
I was wondering if we could incorporate the `join` flag to test this case, with minimal changes to the rest of the code. This is what I came up with:
```
const walk = (workflowGraph, jobName, visitedJobs) => {
    visitedJobs.add(jobName);

    const triggerList = getNextJobs(workflowGraph, { trigger: jobName, prNum: 1 });
    const isJoin = (j1, j2) => !!workflowGraph.edges.find(e => e.src === j1 && e.dest === j2).join;

    // Hit a leaf node, must be good
    if (triggerList.length === 0) {
        return false;
    }

    // Whoops. Can't require yourself.
    if (triggerList.includes(jobName)) {
        return true;
    }

    // Check to see if we visited this job before
    if (triggerList.some(name => visitedJobs.has(name) && !isJoin(jobName, name))) {
        return true;
    }

    // recursively walk starting from the new jobs
    return triggerList.some(name => walk(workflowGraph, name, visitedJobs));
};
```

Reply:
Hmm.. This will work for `join`. But later on, when we support `requires: [~A, ~B]`. This `hasCycle` function won't detect since that won't have a join flag.